openapi.yaml

```diff
@@ -1133,6 +1542,22 @@ components:
       description: Represents a volume within the model
       allOf:
         - $ref: "#/components/schemas/ModelStorageAttributes"
+    Placement:
+      description: Defines a placement directive, which has a scope and a value that is scope-specific.
+      type: object
+      properties:
+        directive:
+          description: Scope-specific placement directive.
+          type: string
+        scope:
+          description: Can be a container type or a machine scope (#).
```

Comment:
You can specify a few provider specific directives also, such as availability zone, all though a zone can be a constraint as well. Check for others which are not constraints.  Some placement directives are specific to bundle and won't be used here.

Reply:
I think you mean the [environment names ](https://github.com/juju/juju/blob/30ba85cdcc5868c64fb9dd9c3f50330d1fa5fac1/core/instance/placement.go#L24), which is not included here (i.e. only containers and existing machine IDs are accepted). Constraint-like things such as the zones are modeled in the `Constraints` schema.